File: Edit_Distance.py
'''
Created on 2014-8-7

Leetcode : Edit_Distance

Given two words word1 and word2, find the minimum number of steps required to

 convert word1 to word2. (each operation is counted as 1 step.)

You have the following 3 operations permitted on a word:

a) Insert a character
b) Delete a character
c) Replace a character

'''

class Solution:
    # @return an integer
    def minDistance(self, word1, word2):
        if not word1:
            return len(word2)
        if not word2:
            return len(word1)
        arr = [ [0 for i in range(len(word1)+1)] for i in range(len(word2)+1) ]
        
        for i in range(len(word1) + 1):
            arr[0][i] = i
        for i in range(len(word2)+1):
            arr[i][0] = i
        for x in range(1,len(word2)+1):
            for y in range(1,len(word1)+1):
                if word1[y-1] == word2[x-1]:
                    dis = 0
                else:
                    dis = 1
                arr[x][y] = min(arr[x-1][y]+1 , arr[x][y-1]+1 , arr[x-1][y-1]+dis) 
        return arr[len(word2)][len(word1)]

test = Solution()
word1 = 'sailn'
word2 =  'failing'
print test.minDistance(word1, word2)

File: Cplusplus/Edit_Distance.cpp
/*
Edit_Distance.cpp

Given two words word1 and word2, find the minimum number of steps required to convert word1 to word2. (each operation is counted as 1 step.)

You have the following 3 operations permitted on a word:

a) Insert a character
b) Delete a character
c) Replace a character
*/

class Solution {
public:
    int minDistance(string word1, string word2) {
        if(word1.size() == 0 || word2.size() == 0)
            return max(word1.size() , word2.size());
        vector<vector<int> > arr(word2.size() + 1 , vector<int>(word1.size() + 1,0));
        for(int i = 1; i < word1.size() +1 ; i++)
        {
            arr[0][i] = i;
        }
        for(int i = 1 ; i < word2.size() + 1 ; i++)
        {
            arr[i][0] = i;
        }
        for(int x = 1 ; x < word2.size() + 1 ; x++)
        {
            for(int y = 1 ; y < word1.size() + 1 ; y++)
            {
                int tmp = min(arr[x-1][y] + 1 , arr[x][y-1] + 1);
                int addStep = 0;
                if(word2[x-1] != word1[y-1])
                    addStep = 1;
                arr[x][y] = min(tmp , arr[x-1][y-1] + addStep);
            }
        }
        return arr[word2.size()][word1.size()];
    }
};
